infrastructure/persistence/user_repository.go

```diff
@@ -1,34 +1,29 @@
 package persistence
 
 import (
+	"github.com/vlpolak/swtgo/cache"
 	"github.com/vlpolak/swtgo/domain/entity"
-	"github.com/vlpolak/swtgo/domain/repository"
 	"gorm.io/gorm"
 )
 
 type UserRepo struct {
 	db *gorm.DB
 }
 
-func (r *UserRepo) FindUser(name string) (*entity.User, map[string]string) {
-	var user entity.User
-	err := r.db.Where("name = ?", name).Take(&user).Error
-	if err != nil {
-		return nil, nil
-	}
-	return &user, nil
-}
-
 func NewUserRepository(db *gorm.DB) *UserRepo {
-	return &UserRepo{db}
+	userRepo := &UserRepo{db}
+	cache.NewLocalCache(userRepo)
+	return userRepo
 }
 
-var _ repository.UserRepository = &UserRepo{}
+func (r *UserRepo) SaveUser(user *entity.User) (*entity.User, error) {
+	err := r.db.Create(&user).Error
+	return user, err
+}
 
-func (r *UserRepo) SaveUser(user *entity.User) (*entity.User, map[string]string) {
-	err := r.db.Create(&user)
-	if err != nil {
-		return nil, nil
-	}
-	return user, nil
+func (r *UserRepo) FindUser(name string) (*entity.User, error) {
+	var user entity.User
+	err := r.db.Where("user_name = ?", name).Take(&user).Error
+	db.Where(entity.User{UserName: name}).FirstOrInit(&user)
```

Comment:
Not sure we need to use `FirstOrInit` here. The most suitable options are
```
err := r.db.Where(&entity.User{UserName: name}).First(&user).Error
```
or one you've provided first
```
err := r.db.Where("user_name = ?", name).Take(&user).Error
```
The difference between `Take` and `First` you can find [here](https://gorm.io/docs/query.html#Retrieving-a-single-object). In theory, `First` more preferable because it always returns first item from ordered set, in opposite to `Take`